### PR TITLE
[Mac] Fix offset mismatch for index() when in a rule containing if() statements

### LINKED ISSUE #1892: [Mac] Keyman on Mojave 10.14.6 emits wrong characters for SIL IPA keyboard
**Background Info**
The updated Keyman as described [here](https://github.com/keymanapp/keyman/issues/1877#issuecomment-514429795) has been install on macOS 10.14.6. The SIL IPA keyboard has also be uninstalled and reinstalled on the OS.

**Describe the bug**

The expected characters are not output. For instance, ɑ, æ and ɐ should be output when pressingː a=, a< and a>.

**To Reproduce**
1. Choose SIL IPA from the keyboard list
2. Go to 'Message' or 'TextEdit' or 'Notes' or 'LibreOffice'
3. Type a=, a<, a>

**Expected behavior**
a= should output ɑ
a< should output æ
a> should output ɐ 

**Screenshots**
Notesː
image

TextEditː
image

Messagesː
image

LibreOfficeː
image

---

**Keyman for macOS**
 - OS: macOS Mojave 10.14.6
 - Keyman version: [12.0 (for testing)](https://drive.google.com/file/d/1HsNAJDlsNGMlmSphVEG4OOmhlgi4NNSa/view?usp=sharing)
 - Target Application: 'Message' or 'TextEdit' or 'Notes' or 'LibreOffice'
   
**Keyboard**
 - Keyboard name: [SIL IPA](https://keyman.com/keyboards/sil_ipa)
 - Keyboard version: 1.7
 - Language name: Phonetic alphabet

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #1892.

This PR includes a second `xLength` function with a subtle difference to solve this bug, because the ObjC version of Keyman Engine is end-of-life (to be replaced with Keyman Core in an upcoming version) and I don't want to waste time refactoring it.

See https://community.software.sil.org/t/sil-ipa-ver-1-7-not-functioning-correctly-on-macos-mojave/2340/5 for original report

### CODE REVIEW

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/Categories/NSString+XString.m [hunk 1]
```diff
@@ -10,6 +10,46 @@
...
+          cx++;
+          i+=3;
+          break;
+        case CODE_INDEX:
+          cx++;
+          i+=4;
+          break;
+        case CODE_IFOPT:
+        case CODE_IFSYSTEMSTORE:
+          cx++;
```
Comment (mcdurdin) [inline mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/Categories/NSString+XString.m]: Here's the one difference to `xLength`.

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/KMEngine.m [hunk 2]
```diff
@@ -848,7 +849,8 @@ - (NSUInteger)contextMatch:(KMCompKey *)cpKey {
                     else {
                         return 0;
                     }
-                    
+
+                    sx++;
```
Comment (mcdurdin) [inline mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/KMEngine.m]: Index offset counter needs to be incremented for `if` statements as well.

File: mac/KeymanEngine4Mac/KeymanEngine4MacTests/KeymanEngineTestsStaticHelperMethods.m [hunk 3]
```diff
@@ -28,4 +28,17 @@ + (KMXFile *)getKmxFileForCipherMusicTests {
     KMXFile *kmxFile = [[KMXFile alloc] initWithFilePath:path];
     return kmxFile;
 }
+
++ (KMXFile *)getKmxFileForSilIpaTests {
+  NSString *path = [[NSBundle bundleForClass:[KeymanEngineTestsStaticHelperMethods class]] pathForResource:@"sil_ipa.kmx" ofType:nil];
```
Comment (mcdurdin) [inline mac/KeymanEngine4Mac/KeymanEngine4MacTests/KeymanEngineTestsStaticHelperMethods.m]: I've added sil_ipa.kmx even though the final test is narrowed down to a single-rule keyboard...

**Comment by keyman-server** [comment]
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/159383:id/kvk/index.html)

**Comment by mcdurdin** [comment]
The Messages issue I think is unrelated, and perhaps needs to be dealt with separately. Am merging :)

**Pull request merged**